### ISSUE TITLE
incorrect commit timestamps

### ISSUE THREAD

**Issue by nustiueudinastea** [dolthub/dolt#6700]
Dear Dolt team,

I am using Dolt as an embedded DB in a Go codebase (with some modifications) and I have noticed peculiar behaviour related to the timestamps of the commits. The following is a picture that shows the `dolt_commits` tables without DATE ordering.

<img width="889" alt="Screenshot 2023-09-21 at 12 41 58" src="https://github.com/dolthub/dolt/assets/588327/0f317547-aac5-43d1-84f8-7185ce991d10">

As far as I can see there are 2 issues:
1. The first commit with message "Initialize data repository", has a commit date 1 millisecond after the second commit.
  * In my code the second commit ("Initialize doltswarmdemo") happens time wise after the 1st commit
  * Dolt log shows the correct ordering but somehow the timestamps are wrong.
  * These 2 first commits are done doing an init phase of the program after which the program quits
3. The rest of the commits have the same timestamps
  * all these commits are done in a second instantiations of the program
  * time wise, these happen each 15 seconds apart, in a loop that goes on forever

If it's useful, I could open up the repos for my code on Github. I want to also mention that it's possible that this happens because of how I initialise the Dolt database in my code.

The insert and commit pattern looks like this:

```
        queryString := fmt.Sprintf("INSERT INTO %s (id, name) VALUES ('%s', '%s');", tableName, uid.String(), data)

	tx, err := dbi.Begin()
	if err != nil {
		return fmt.Errorf("failed to start transaction: %w", err)
	}

	defer tx.Rollback()

	_, err = tx.Exec(queryString)
	if err != nil {
		return fmt.Errorf("failed to save record: %w", err)
	}

	// add
	_, err = tx.Exec(`CALL DOLT_ADD('-A');`)
	if err != nil {
		return fmt.Errorf("failed to commit table: %w", err)
	}

	// commit
	_, err = tx.Exec(`CALL DOLT_COMMIT('-m', 'Inserted time', '--author', 'Alex Giurgiu <alex@giurgiu.io>');`)
	if err != nil {
		return fmt.Errorf("failed to commit table: %w", err)
	}

	err = tx.Commit()
	if err != nil {
		return fmt.Errorf("failed to commit insert transaction: %w", err)
	}
```

In my code I managed to work around this issue by adding my own commit `--date` time stamp, using RFC3339Nano format. This seems to give the amount of granularity I need.

**Comment by fulghum** [comment]
Hi @nustiueudinastea, thanks for reporting this. Glad to hear you have a temporary workaround while we dig in and see what's going on. 

It sounds like there's a session timestamp that's being held onto and not being updated properly when a new query executes. I don't see anything wrong with the code you provided above. Can you share more details about how you're embedding Dolt in your Go application and connecting? Are you using the [dolthub/driver](https://github.com/dolthub/driver) code to connect? 

I'll set up a quick demo with dolthub/driver and see if I can reproduce the behavior.

**Comment by nustiueudinastea** [comment]
Let me publish the repos that I was working on as I did some modifications to dolt in order to support my use case. Will get back to you

**Comment by fulghum** [comment]
Sounds great! I think I may have a repro with the dolthub/driver sample code. I'll keep digging in there...

**Comment by nustiueudinastea** [comment]
I published the repos. The actual place where I do the db init is here: https://github.com/nustiueudinastea/doltswarm/blob/614c6c61cdb4b51dc088904fa76d8235f893d3fb/db.go#L129

The other repo where I import doltswarm and do commits is here: https://github.com/nustiueudinastea/distributeddolt

**Comment by fulghum** [comment]
Thanks @nustiueudinastea, I took a quick look at the db initialization code you linked to, and it lines up with what I found in dolthub/driver. 

When a `doltStmt` is created, we are always using the same `gmsCtx` instance from the `doltConn`, without ever updating the `QueryTime` field:
https://github.com/dolthub/driver/blob/ca79c86378d67263b78cd10c2bbe7f69d6ab2bd9/conn.go#L81-L80

That `QueryTime` field is where `dolt_commit()` gets its date from and once I started calling `SetQueryTime(time.Now())` in the code above that creates each new `doltStmt`, the dolt commits get the correct timestamps.

I need to poke around a little more in this package, make sure this is the right way to fix this, and dig into how we're testing this package, but I should be able to get a PR opened shortly and then get the fix out for you.

**Comment by nustiueudinastea** [comment]
Awesome! Will test it as well once you have PR. Thanks so much!

**Comment by fulghum** [comment]
@nustiueudinastea – The fix is merged into main for `dolthub/driver`. After you get the latest version from main ([af8c3cf](https://github.com/dolthub/driver/commit/af8c3cfb8ca4ed905ced7f8b7c1864720eb12c1c)), you should see new `dolt_commits` use the correct timestamp. 

Thanks for reporting this issue so we could fix it! 🙏  I'll leave this open a little longer in case you hit any surprises, but I think this should do the trick!

**Comment by nustiueudinastea** [comment]
Hey @fulghum , will be able to test this tomorrow. Much appreciated!

**Comment by nustiueudinastea** [comment]
Problem has been fixed, thank you so much for your prompt work!

<img width="858" alt="Screenshot 2023-09-22 at 11 02 45" src="https://github.com/dolthub/dolt/assets/588327/cfd5c827-bc51-4dd8-b7a8-2eb116d02d22">

I have updated my forks to the latest code from your repos, and now I see the the timestamps don't contain millisecond accuracy anymore. Or at least, the select doesn't show the millisecond. Was this on purpose? Closing this issue.
